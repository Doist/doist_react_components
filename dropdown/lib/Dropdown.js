(function() {
  var Dropdown, OptionItem, Options, React, Trigger;

  React = require('react');

  Dropdown = React.createClass({
    getInitialState: function() {
      return {
        show_options: false
      };
    },
    _toggle: function(event) {
      if (event) {
        event.stopPropagation();
        event.preventDefault();
      }
      if (this.state.show_options && this.props.onDropdownClose) {
        this.props.onDropdownClose();
      }
      if (!this.state.show_options && this.props.onDropdownOpen) {
        this.props.onDropdownOpen();
      }
      return this.setState({
        show_options: !this.state.show_options
      });
    },
    _renderMenuTrigger: function() {
      var menu_trigger;
      menu_trigger = null;
      React.Children.forEach(this.props.children, (function(_this) {
        return function(child) {
          if (child.type === Trigger) {
            return menu_trigger = React.cloneElement(child, {
              onClick: _this._toggle
            });
          }
        };
      })(this));
      return menu_trigger;
    },
    _renderMenuOptions: function() {
      var menu_options;
      if (!this.state.show_options) {
        return false;
      }
      menu_options = null;
      React.Children.forEach(this.props.children, (function(_this) {
        return function(child) {
          if (child.type === Options) {
            return menu_options = React.createElement("div", {
              "className": "o-dropdown__options_renderer"
            }, React.cloneElement(child, {
              hideOptions: _this._toggle,
              on_top: _this.props.options_on_top
            }));
          }
        };
      })(this));
      return menu_options;
    },
    render: function() {
      return React.createElement("div", {
        "className": "o-dropdown"
      }, (this.props.options_on_top === true ? this._renderMenuOptions() : void 0), this._renderMenuTrigger(), (this.props.options_on_top !== true ? this._renderMenuOptions() : void 0));
    }
  });

  Trigger = React.createClass({
    render: function() {
      return React.createElement("div", {
        "className": "o-dropdown__trigger",
        "onClick": this.props.onClick
      }, this.props.children);
    }
  });

  Options = React.createClass({
    _handleClick: function(event) {
      return this.props.hideOptions();
    },
    componentWillMount: function() {
      return document.addEventListener("click", this._handleClick, false);
    },
    componentWillUnmount: function() {
      return document.removeEventListener("click", this._handleClick, false);
    },
    render: function() {
      var className;
      className = "o-dropdown__options";
      if (this.props.on_top) {
        className += " o-dropdown__options--on_top";
      }
      if (this.props.align === "right") {
        className += " o-dropdown__options--align_right";
      }
      return React.createElement("div", {
        "className": className
      }, React.createElement("ul", null, this.props.children));
    }
  });

  OptionItem = React.createClass({
    render: function() {
      return React.createElement("li", null, this.props.children);
    }
  });

  Dropdown.Trigger = Trigger;

  Dropdown.Options = Options;

  Dropdown.OptionItem = OptionItem;

  module.exports = Dropdown;

}).call(this);
