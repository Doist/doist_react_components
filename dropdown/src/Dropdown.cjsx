React = require('react')

Dropdown = React.createClass

    getInitialState: ->
        return { show_options: false }

    _toggle: ->
        if @state.show_options && @props.onDropdownClose
            @props.onDropdownClose()
        if !@state.show_options && @props.onDropdownOpen
            @props.onDropdownOpen()
        @setState { show_options: !@state.show_options }

    _renderMenuTrigger: ->
        menu_trigger = null
        React.Children.forEach(@props.children, (child) =>
            if child.type == Trigger
                menu_trigger = React.cloneElement(child, {onClick: @_toggle})
        )
        return menu_trigger

    _renderMenuOptions: ->
        if !@state.show_options
            return false

        menu_options = null
        React.Children.forEach @props.children, (child) =>
            if child.type == Options
                menu_options =
                    <div className = "o-dropdown__options_renderer">
                        {React.cloneElement(child, {hideOptions: @_toggle, on_top: @props.options_on_top})}
                    </div>

        return menu_options

    render: ->
        <div className = "o-dropdown">
            { @_renderMenuOptions() if @props.options_on_top == true }
            { @_renderMenuTrigger() }
            { @_renderMenuOptions() if @props.options_on_top != true }
        </div>

Trigger = React.createClass
    render: ->
        <div className = "o-dropdown__trigger", onClick = @props.onClick>
            { @props.children }
        </div>

Options = React.createClass
    _handleClick: (event) ->
        if @getDOMNode().contains(event.target)
            return
        else
            @props.hideOptions()

    # when options are displaying will be listening for clicks.
    # will hide options if click outside.
    componentWillMount: ->
        document.addEventListener("click", @_handleClick, false)

    componentWillUnmount: ->
        document.removeEventListener("click", @_handleClick, false)

    render: ->
        className = "o-dropdown__options"
        className += " o-dropdown__options--on_top" if @props.on_top
        className += " o-dropdown__options--align_right" if @props.align == "right"

        <div className = className>
            <ul>
                { @props.children }
            </ul>
        </div>

OptionItem = React.createClass
    render: ->
        <li>
            { @props.children }
        </li>

Dropdown.Trigger = Trigger
Dropdown.Options = Options
Dropdown.OptionItem = OptionItem
module.exports = Dropdown
