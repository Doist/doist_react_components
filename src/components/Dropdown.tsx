import './styles/dropdown.less'

import React from 'react'
import ReactDOM from 'react-dom'
import classNames from 'classnames'

type BoxProps = {
    onShowBody?: () => void
    onHideBody?: () => void
    allowBodyInteractions?: boolean
    top?: boolean
    right?: boolean
    scrolling_parent?: string
    children?: [
        React.ReactElement<TriggerProps>,
        React.ReactElement<{}> | ((props: {}) => JSX.Element)
    ]
    className?: string
}

type BoxState = {
    top: boolean
    show_body: boolean
}

class Box extends React.Component<BoxProps, BoxState> {
    public static displayName: string

    constructor(props: BoxProps, context: React.Context<any>) {
        super(props, context)
        this.state = {
            //eslint-disable-next-line @typescript-eslint/camelcase
            show_body: false,
            top: props.top || false,
        }

        this._handleClickOutside = this._handleClickOutside.bind(this)
        this._setPosition = this._setPosition.bind(this)
        this._toggleShowBody = this._toggleShowBody.bind(this)
        this._timeout = undefined
    }

    componentWillUnmount() {
        document.removeEventListener('click', this._handleClickOutside, true)
        if (this._timeout) {
            clearTimeout(this._timeout)
        }
    }
    _timeout?: number

    _handleClickOutside(event: MouseEvent) {
        //eslint-disable-next-line @typescript-eslint/camelcase
        const dropdown_dom_node = ReactDOM.findDOMNode(this)

        if (
            //eslint-disable-next-line @typescript-eslint/camelcase
            dropdown_dom_node &&
            //eslint-disable-next-line @typescript-eslint/camelcase
            !dropdown_dom_node.contains(event.target as Node)
        )
            this._toggleShowBody()
        else if (!this.props.allowBodyInteractions) {
            // won't close when body interactions are allowed
            this._timeout = setTimeout(() => {
                if (this.state.show_body) {
                    this._toggleShowBody()
                }
            }, 100)
        }
    }

    _toggleShowBody() {
        if (!this.state.show_body) {
            // will show
            if (this.props.onShowBody) this.props.onShowBody()
            document.addEventListener('click', this._handleClickOutside, true)
        } else {
            // will hide
            if (this.props.onHideBody) this.props.onHideBody()
            document.removeEventListener(
                'click',
                this._handleClickOutside,
                true
            )
        }

        this.setState({
            //eslint-disable-next-line @typescript-eslint/camelcase
            show_body: !this.state.show_body,
        })
    }

    _getTriggerComponent() {
        const _trigger = this.props.children && this.props.children[0]
        return _trigger
            ? React.cloneElement(_trigger, { onClick: this._toggleShowBody })
            : undefined
    }

    // https://facebook.github.io/react/docs/refs-and-the-dom.html#exposing-dom-refs-to-parent-components
    _setPosition(body: HTMLElement) {
        if (body) {
            //eslint-disable-next-line @typescript-eslint/camelcase
            const scrolling_parent = document.getElementById(
                this.props.scrolling_parent ? this.props.scrolling_parent : ''
            )

            /* eslint-disable @typescript-eslint/camelcase */
            if (scrolling_parent) {
                const dropdown = ReactDOM.findDOMNode(this)
                if (!dropdown) {
                    return
                }
                const dropdown_vertical_position = /** @type {HTMLElement} */ (ReactDOM.findDOMNode(
                    this
                ) as HTMLElement).offsetTop
                const dropdown_trigger = (dropdown as Element).querySelector(
                    '.trigger'
                )
                if (!dropdown_trigger) {
                    return
                }
                const dropdown_trigger_height =
                    /** @type {Element} */ dropdown_trigger.clientHeight
                const dropdown_body_height = body.clientHeight

                const scrolling_parent_height = scrolling_parent.clientHeight
                const scrolling_parent_offset = scrolling_parent.scrollTop

                const bottom_offset =
                    scrolling_parent_height +
                    scrolling_parent_offset -
                    dropdown_vertical_position -
                    dropdown_trigger_height

                const top = bottom_offset < dropdown_body_height

                if (top !== this.state.top) {
                    this.setState({ top })
                }
            }
            /* eslint-enable @typescript-eslint/camelcase */
        }
    }

    _getBodyComponent() {
        if (!this.state.show_body) {
            return null
        }
        const { top } = this.state
        const { right = false, children } = this.props
        const props = { top, right, setPosition: this._setPosition }

        /* eslint-disable @typescript-eslint/camelcase */
        const class_name = classNames({
            body_wrapper: true,
            with_arrow: true,
            top: top,
            bottom: !top,
        })
        /* eslint-enable @typescript-eslint/camelcase */

        const body = children && children[1]

        const contentMarkup =
            typeof body === 'function'
                ? body(props)
                : body
                ? React.cloneElement(body, props)
                : undefined
        return (
            //eslint-disable-next-line @typescript-eslint/camelcase
            <div className={class_name} style={{ position: 'relative' }}>
                {contentMarkup}
            </div>
        )
    }

    render() {
        const className = classNames('reactist_dropdown', this.props.className)
        const { top } = this.state

        return (
            <div style={{ display: 'inline-block' }} className={className}>
                {top && this._getBodyComponent()}
                {this._getTriggerComponent()}
                {!top && this._getBodyComponent()}
            </div>
        )
    }
}
Box.displayName = 'Dropdown.Box'

type TriggerProps = {
    onClick?: (event?: React.MouseEvent) => void
}

class Trigger extends React.Component<TriggerProps> {
    public static displayName: string

    constructor(props: TriggerProps, context: React.Context<unknown>) {
        super(props, context)
        this._onClick = this._onClick.bind(this)
    }

    _onClick(event: React.MouseEvent) {
        event.preventDefault()
        event.stopPropagation()
        if (this.props.onClick) this.props.onClick(event)
    }

    render() {
        const style = {
            display: 'block',
        }

        return (
            <div style={style} className="trigger" onClick={this._onClick}>
                {this.props.children}
            </div>
        )
    }
}
Trigger.displayName = 'Dropdown.Trigger'

class Body extends React.Component<any, any, any> {
    public static displayName: string

    render() {
        const style: React.CSSProperties = {
            position: 'absolute',
            right: 0,
            top: 0,
        }

        if (this.props.top) {
            style.top = 'auto'
            style.bottom = 0
        }

        if (this.props.right) {
            style.right = 'auto'
            style.left = 0
        }

        return (
            <div
                ref={this.props.setPosition}
                style={style}
                className="body"
                id="reactist-dropdown-body"
            >
                {this.props.children}
            </div>
        )
    }
}
Body.displayName = 'Dropdown.Body'

const Dropdown = {
    Box,
    Trigger,
    Body,
}

export default Dropdown
