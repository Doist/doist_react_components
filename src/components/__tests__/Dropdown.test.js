import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import Dropdown from '../Dropdown'

describe('Dropdown.Trigger', () => {
    it('renders empty without crashing', () => {
        const trigger = shallow(<Dropdown.Trigger />)
        expect(toJson(trigger)).toMatchSnapshot()
    })

    it('renders all children without crashing', () => {
        const simpleTrigger = shallow(
            <Dropdown.Trigger>Trigger Content</Dropdown.Trigger>
        )
        expect(toJson(simpleTrigger)).toMatchSnapshot()

        const complexTrigger = shallow(
            <Dropdown.Trigger>
                <div>
                    <img src="" alt="" />
                    Some additional text as well
                </div>
            </Dropdown.Trigger>
        )
        expect(toJson(complexTrigger)).toMatchSnapshot()
    })

    it('calls onClick callback', () => {
        const onClickSpy = jest.fn()
        const trigger = shallow(<Dropdown.Trigger onClick={onClickSpy} />)

        trigger.simulate('click', {
            preventDefault: jest.fn(),
            stopPropagation: jest.fn()
        })

        expect(onClickSpy).toHaveBeenCalledTimes(1)
    })
})
