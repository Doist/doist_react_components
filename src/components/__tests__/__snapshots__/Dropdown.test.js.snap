// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown.Trigger renders all children without crashing 1`] = `
<div
  className="trigger"
  onClick={[Function]}
  style={
    Object {
      "display": "block",
    }
  }
>
  Trigger Content
</div>
`;

exports[`Dropdown.Trigger renders all children without crashing 2`] = `
<div
  className="trigger"
  onClick={[Function]}
  style={
    Object {
      "display": "block",
    }
  }
>
  <div>
    <img
      alt=""
      src=""
    />
    Some additional text as well
  </div>
</div>
`;

exports[`Dropdown.Trigger renders empty without crashing 1`] = `
<div
  className="trigger"
  onClick={[Function]}
  style={
    Object {
      "display": "block",
    }
  }
/>
`;
